libsoftwarecontainer/src/gateway/devicenode/devicenode.cpp

```diff
@@ -0,0 +1,138 @@
+ * SOFTWARE.
+ *
+ * For further information see LICENSE
+ */
+
+#include <sys/stat.h>
+
+#include "devicenode.h"
+#include "functionjob.h"
+
+namespace softwarecontainer {
+
+Device::Device(std::string _name, int _mode) :
+        name(_name),
+        mode(_mode),
+        isConfigured(false)
+{ }
+
+Device::Device(const Device &dev) :
+        name(dev.name),
+        mode(dev.mode),
+        isConfigured(false)
+{ }
+
+bool Device::parse(const json_t *element)
+{
+
+    name = "";
+    if (!JSONParser::read(element, "name", name)) {
+        log_error() << "Key \"name\" missing or not a string in json configuration";
+        return false;
+    }
+
+    mode = -1;
+
+    // Mode is optional, i.e. do not do anything if it is not specified.
+    if (nullptr != json_object_get(element, "mode")) {
+        const bool modeParses = JSONParser::read(element, "mode", mode);
+
+        if (!modeParses) {
+            log_error() << "Mode specified with bad format";
+            return false;
+        }
+    }
+
+    return true;
+}
+
+bool Device::activate(std::shared_ptr<ContainerAbstractInterface> container)
+{
+    if (!isConfigured) {
+        log_info() << "Mapping device " << name;
+        // Mount device in container
+        if (!container->mountDevice(name)) {
+            log_error() << "Unable to mount device " << name;
+            return false;
+        }
+
+        // If mode is specified, try to set mode for the mounted device.
+        if (mode != -1) {
+            FunctionJob job(container, [&] () {
+                return chmod(name.c_str(), mode);
+            });
+            job.start();
+            job.wait();
+            if (job.isError()) {
+                log_error() << "Could not 'chmod " << mode
+                            << "' the mounted device " << name;
+                return false;
+            }
+        }
+        isConfigured = true;
+    }
+    return true;
+}
+
+void Device::calculateDeviceMode(const int appliedMode)
+{
+    int modeResult = mode;
+
+    if (mode != appliedMode) {
+        //apply more permissive option for owner
+        ((appliedMode / 100) >= (mode / 100)) ?
+                modeResult = (appliedMode / 100) * 100 : modeResult = (mode / 100) * 100;
+        //apply more permissive option for group
+        (((appliedMode / 10) % 10) >= ((mode / 10) % 10)) ?
+                modeResult += (((appliedMode / 10) % 10) * 10) : modeResult += (((mode / 10) % 10) * 10);
+        //apply more permissive option for others
+        ((appliedMode % 10) >= (mode % 10)) ?
+                modeResult += (appliedMode % 10) :  modeResult += (mode % 10);
+
+    }
+
+    if (modeResult > mode) {
+        isConfigured = false;
+        mode = modeResult;
+    }
+}
+
+const std::string Device::getName()
+{
+    return name;
+}
+
+int Device::getMode()
+{
+    return mode;
+}
+
+bool Device::getIsconfigured ()
+{
+    return isConfigured;
+}
```

Comment:
I would prefer if the getters had the name of the variable, so mode(), name() etc.